src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
     SPELL_MEMORY_VEZAX          = 66714,
 
     // Memory
+    SPELL_SHADOWFORM            = 41408,
+    SPELL_SPAWN_VISUAL          = 66675,
     SPELL_OLD_WOUNDS            = 66620,
-    SPELL_OLD_WOUNDS_H          = 67679,
     SPELL_SHADOWS_PAST          = 66619,
-    SPELL_SHADOWS_PAST_H        = 67678,
-    SPELL_WAKING_NIGHTMARE      = 66552,
-    SPELL_WAKING_NIGHTMARE_H    = 67677
+    SPELL_WAKING_NIGHTMARE      = 66552
 };
 
-class OrientationCheck : public std::unary_function<Unit*, bool>
+enum PointMovement
 {
-    public:
-        explicit OrientationCheck(Unit* _caster) : caster(_caster) { }
-        bool operator()(WorldObject* object)
-        {
-            return !object->isInFront(caster, 2.5f) || !object->IsWithinDist(caster, 40.0f);
-        }
-
-    private:
-        Unit* caster;
+    POINT_PREFIGHT              = 0,
+    POINT_PREPARE,
+    POINT_DESPAWN
 };
 
-class spell_eadric_radiance : public SpellScriptLoader
+struct argent_challenge_baseAI : public BossAI
 {
-    public:
-        spell_eadric_radiance() : SpellScriptLoader("spell_eadric_radiance") { }
-        class spell_eadric_radiance_SpellScript : public SpellScript
-        {
-            PrepareSpellScript(spell_eadric_radiance_SpellScript);
+    argent_challenge_baseAI(Creature* creature) : BossAI(creature, DATA_ARGENT_CHALLENGE)
+    {
+        Initialize();
+        creature->SetReactState(REACT_DEFENSIVE);
+    }
 
-            void FilterTargets(std::list<WorldObject*>& unitList)
-            {
-                unitList.remove_if(OrientationCheck(GetCaster()));
-            }
+    virtual void Initialize() { }
 
-            void Register() override
-            {
-                OnObjectAreaTargetSelect += SpellObjectAreaTargetSelectFn(spell_eadric_radiance_SpellScript::FilterTargets, EFFECT_0, TARGET_UNIT_SRC_AREA_ENEMY);
-                OnObjectAreaTargetSelect += SpellObjectAreaTargetSelectFn(spell_eadric_radiance_SpellScript::FilterTargets, EFFECT_1, TARGET_UNIT_SRC_AREA_ENEMY);
-            }
-        };
+    virtual void OnReset()
+    {
+        _Reset();
+    }
+
+    void EnterEvadeMode(EvadeReason /*why*/) final override
+    {
+        if (instance->GetBossState(DATA_ARGENT_CHALLENGE) != SPECIAL &&
+            instance->GetBossState(DATA_ARGENT_CHALLENGE) != DONE)
+            BossAI::EnterEvadeMode();
+    }
+
+    void JustReachedHome() final override
+    {
+        if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == SPECIAL)
+            events.ScheduleEvent(EVENT_CHALLENGE_DONE, 4 * IN_MILLISECONDS);
+        else
+            me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_STATE_READY1H);
+    }
+
+    void Reset() final override
+    {
+        if (instance->GetBossState(DATA_ARGENT_CHALLENGE) != SPECIAL)
+            OnReset();
+    }
 
-        SpellScript* GetSpellScript() const override
+    void AttackStart(Unit* who) final override
+    {
```

Comment:
I _really_ don't like this sort of `AttackStart` replacement. React states instead, please.